### PR TITLE
Convert some async TestCase code to coroutines 

### PR DESCRIPTION
**Pull request by p12tic**
This PR cleans up some asynchronous TestCase code. Previous code was relatively hard to follow due to it relying on callbacks based implementation, which forced the logic to be split across many small functions. The new code puts everything into regular code flow across several functions and uses standard `try...except` and `try...finally` for error handling.

The code would be even simplier if `IReporter.addError` supported regular exceptions, but this is for a future.

The PR has been split across many small commits so that it is possible to bisect the changes in the future if there are any bugs that may be caused by this PR. Also, review should be easier, as the changes in isolation are almost trivial.

### CODE REVIEW

**Comment by p12tic** [comment]
please review

File: src/twisted/trial/_asynctest.py [hunk 1]
```diff
@@ -131,43 +131,42 @@ def onTimeout(d):
...
-        d.addCallbacks(
-            self._cbDeferTestMethod,
-            self._ebDeferTestMethod,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        d.addBoth(self.deferRunCleanups, result)
-        d.addBoth(self.deferTearDown, result)
-        return d
+    async def deferSetUpNoCleanup(self, result):
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: Just asking. Can we make this private   ?

File: src/twisted/trial/_asynctest.py [hunk 2]
```diff
@@ -131,43 +131,42 @@ def onTimeout(d):
...
-        d = self._run("setUp", result)
-        d.addCallbacks(
-            self.deferTestMethod,
-            self._ebDeferSetUp,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        return d
+    async def deferSetUp(self, result):
+        self._runNeedsTearDown = False
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: Can we also initialized it in `__init__` ?

I think that this needs documentation as part of TestCase docstring. 

Even if this is private, we should add a few words about why it exists

File: src/twisted/trial/_asynctest.py [hunk 3]
```diff
@@ -131,43 +131,42 @@ def onTimeout(d):
...
-    def _ebDeferSetUp(self, failure, result):
-        if failure.check(SkipTest):
-            result.addSkip(self, self._getSkipReason(self.setUp, failure.value))
-        else:
-            result.addError(self, failure)
-            if failure.check(KeyboardInterrupt):
-                result.stop()
-        return self.deferRunCleanups(None, result)
-
-    def deferTestMethod(self, ignored, result):
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: Was this method removed ?

This looks like it's part of the public API

-------

I hope that we should keep the method and:
* implement it via `deferSetUpNoCleanup` (or the private version of deferSetUpNoCleanup)
* add a warning that it is deprecated

I am also ok, just to add a comment that the method is deprecated and add  `pragma: nocover` to make coverage happy

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the cleanup and sorry for the delay in reviewing.

The changes looks good.

My only concern is about removign the public method.

File: src/twisted/trial/_asynctest.py (hunk 1 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Sure.

File: src/twisted/trial/_asynctest.py (hunk 3 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: I assumed that `deferSetUp`, `deferTestMethod` and `deferTearDown` are effectively private anyway. There is no case where they would be useful to override. Also, their API is not "public ready" (i.e. the `ignored` parameter) - it seems as if they were some kind of internal thing that simply didn't get the `_` prefix.

Additionally, I don't think we can add a deprecation warning to these functions easily. We need to detect not when they are called, but when they are overridden. This means that all the current logic must still use these functions. This constrains the implementation a lot and means that this PR cannot go in until these functions can be removed.

Finally, I did search for these methods in GitHub and there's no public code that overrides them.

Given the above - small risk of breaking code for someone and relatively high costs for going through the deprecation dance I would suggest to have an exception for this change and allow to remove the functions without deprecation period.

I will need to update the release notes accordingly.

If we go through the path of removing the functions outright, it probably makes sense to have `assert not hasattr(self, 'deferTestMethod'), "overriding deferTestMethod is not supported"` just before setUpClass is executed, so that if there is someone who does use the functions after all, he won't get silent failures.

Please let me know what you think.

Comment (adiroiban) [reply src/twisted/trial/_asynctest.py]: Thanks for the info.
This is something that I have missed during the initial review.

-------


> There is no case where they would be useful to override.

This might be true.
But I still think that we should try to follow the "best practices" and handle this as removing public API.

We need to add a `12375.removal` newsfragment and inform Twisted trial users that this change was made.

And if we are doing the removal, I prefer to have the new names as private names... to reduce the need for any future communication.

And it will also need an email to the mailing list to inform about this change and ask for an exception.
I can help with mailing list communication

--------

As there are not many developers dedicating time to maintain `trial` I am +1 for making an exception here and removing these public methods without a deprecation warning.

SO I am +1 for approving the exception.

File: src/twisted/trial/_asynctest.py [hunk 4]
```diff
@@ -131,43 +131,42 @@ def onTimeout(d):
...
-    def deferSetUp(self, ignored, result):
-        d = self._run("setUp", result)
-        d.addCallbacks(
-            self.deferTestMethod,
-            self._ebDeferSetUp,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        return d
+    async def deferSetUp(self, result):
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: As we change the signature and the return value of this method, I think is best to make it private.

```suggestion
    async def _deferSetUp(self, result):
```

File: src/twisted/trial/_asynctest.py [hunk 5]
```diff
@@ -131,43 +131,42 @@ def onTimeout(d):
...
+        self._runNeedsTearDown = False
+        try:
+            try:
+                await self.deferSetUpNoCleanup(result)
+            finally:
+                await self.deferRunCleanups(None, result)
+        finally:
+            if self._runNeedsTearDown:
+                await self.deferTearDown(None, result)
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: To simplify the code
```suggestion
                await self._deferTearDown(result)
```

File: src/twisted/trial/_asynctest.py [hunk 6]
```diff
@@ -185,16 +184,16 @@ def _ebDeferTestMethod(self, f, result):
...
 
-    def deferTearDown(self, ignored, result):
-        d = self._run("tearDown", result)
-        d.addErrback(self._ebDeferTearDown, result)
-        return d
-
-    def _ebDeferTearDown(self, failure, result):
-        result.addError(self, failure)
-        if failure.check(KeyboardInterrupt):
+    async def deferTearDown(self, ignored, result):
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: Since we are changing the return value, 
I think that is best to make the "new method" private.

```suggestion
    async def _deferTearDown(self, result):
```

File: src/twisted/trial/_asynctest.py (hunk 3 above)
Comment (glyph) [reply src/twisted/trial/_asynctest.py]: > Given the above - small risk of breaking code for someone and relatively high costs for going through the deprecation dance I would suggest to have an exception for this change and allow to remove the functions without deprecation period.

I'm inclined to agree with your rationale for the exception, but please [follow the already-documented process for exceptions](https://docs.twisted.org/en/stable/development/compatibility-policy.html#procedure-for-exceptions-to-this-policy) :)

Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Sure, I wasn't aware that I'm removing public API, as soon as Adi raised this the PR became totally wrong.

Comment (adiroiban) [reply src/twisted/trial/_asynctest.py]: I can help with sending the email to announce the process exceptions.

@p12tic as long as you think want to land this PR, I think that we can do it... it's just that the PR need a `.removal` newsfragment not a `.misc`

Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: @adiroiban I opened a separate PR for the removal https://github.com/twisted/twisted/pull/12389

**Comment by p12tic** [comment]
With rejection of https://github.com/twisted/twisted/issues/12387 I don't think this PR has benefits, therefore I'm closing it.

**Comment by twm** [comment]
I wish we had some benchmarks here, because I wonder if this speed up Trial tests (which would be very much worthwhile!).

**Comment by p12tic** [comment]
Benchmarks have been added in #12393. On my machine I see improvement of around 2%, so probably it makes sense to land this PR still.

File: src/twisted/trial/_asynctest.py (hunk 1 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Fixed now.

File: src/twisted/trial/_asynctest.py (hunk 3 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Addressed as part of https://github.com/twisted/twisted/pull/12389.

File: src/twisted/trial/_asynctest.py (hunk 6 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Fixed now.

File: src/twisted/trial/_asynctest.py (hunk 5 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Fixed now.

File: src/twisted/trial/_asynctest.py (hunk 4 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Fixed now.

File: src/twisted/trial/_asynctest.py (hunk 2 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Fixed now.

**Comment by p12tic** [comment]
@adiroiban I have addressed all your comments.

File: src/twisted/trial/_asynctest.py [hunk 7]
```diff
@@ -134,43 +138,43 @@ def onTimeout(d):
...
-        d.addCallbacks(
-            self._cbDeferTestMethod,
-            self._ebDeferTestMethod,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        d.addBoth(self._deferRunCleanups, result)
-        d.addBoth(self._deferTearDown, result)
-        return d
+    async def _deferSetUpNoCleanup(self, result):
```
Comment (adiroiban) [inline src/twisted/trial/_asynctest.py]: Maybe this is a better name


```suggestion
    async def _deferSetUpAndRun(self, result):
        """
        Execute the setUp and run part of a test.

        Teardown and cleanups are not executed.
        """
```

**Review by adiroiban — APPROVED** [review]
Looks good. Thanks for the refactoring.

Only a minor comment.

File: src/twisted/trial/_asynctest.py (hunk 7 above)
Comment (p12tic) [reply src/twisted/trial/_asynctest.py]: Applied now, thanks.
